h2o-core/src/main/java/hex/segments/SegmentModelsBuilder.java

```diff
@@ -116,24 +127,51 @@ private Frame reorderColumns(Key<Frame> key) {
   }
 
   private static class MultiNodeRunner extends MRTask<MultiNodeRunner> {
-    LocalSequentialSegmentModelsBuilder _builder;
-    SegmentModels _segment_models;
+    final LocalSequentialSegmentModelsBuilder _builder;
+    final SegmentModels _segment_models;
+    final int _parallelism;
+
+    // OUT
+    LocalSequentialSegmentModelsBuilder.SegmentModelsStats _stats;
 
-    private MultiNodeRunner(LocalSequentialSegmentModelsBuilder builder, SegmentModels segmentModels) {
+    private MultiNodeRunner(LocalSequentialSegmentModelsBuilder builder, SegmentModels segmentModels, int parallelism) {
       _builder = builder;
       _segment_models = segmentModels;
+      _parallelism = parallelism;
     }
 
     @Override
     protected void setupLocal() {
-      _builder.buildModels(_segment_models);
+      if (_parallelism == 1) {
+        _stats = _builder.buildModels(_segment_models);
+      } else {
+        List<Callable<SegmentModelsStats>> runnables = Stream.<Callable<SegmentModelsStats>>generate(
+                () -> (() -> _builder.clone().buildModels(_segment_models))
+        ).limit(_parallelism).collect(Collectors.toList());
+        try {
+          ExecutorService executor = Executors.newFixedThreadPool(_parallelism);
+          _stats = new SegmentModelsStats();
+          for (Future<SegmentModelsStats> f : executor.invokeAll(runnables)) {
+            _stats.reduce(f.get());
+          }
+        } catch (ExecutionException | InterruptedException e) {
+          throw new RuntimeException("Failed to build segment-models", e);
+        }
```

Comment:
This can be written as a stream in it's whole beauty. Optional of course. 

```suggestion
        ExecutorService executor = Executors.newFixedThreadPool(_parallelism);
        _stats = Stream.<Callable<SegmentModelsStats>>generate(
                () -> (() -> _builder.clone().buildModels(_segment_models)))
                .limit(_parallelism)
                .map(callable -> executor.submit(callable))
                .map(future -> {
                  try {
                    return future.get();
                  } catch (ExecutionException | InterruptedException e) {
                    throw new RuntimeException("Failed to build segment-models", e);
                  }
                })
                .reduce((a, b) -> {
                  a.reduce(b);
                  return a;
                }).get();
```

Reply:
nice trick with using submit instead of invokeAll to get rid of the outer InterruptedException

thank you @Pscheidl